Editor/MenuActions/Selection/SelectPathFaces.cs

```diff
@@ -0,0 +1,157 @@
+﻿using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.ProBuilder;
+using System.Linq;
+
+namespace UnityEditor.ProBuilder.Actions
+{
+    public static class SelectPathFaces
+    {
+        private static int[] lastPredecessors;
+        private static int lastStart;
+        private static ProBuilderMesh lastMesh;
+        private static List<WingedEdge> lastWings;
+        private static Dictionary<Face, int> lastFacesIndex = new Dictionary<Face, int>();
+
+        /// <summary>
+        /// Calculates the indexes of all faces in the shortest path between start and end
+        /// </summary>
+        /// <param name="start">The index of the starting face</param>
+        /// <param name="end">The index of the ending face</param>
+        /// <param name="mesh">The mesh of the object</param>
+        /// <returns>The indexes of all faces </returns>
+        public static List<int> GetPath(int start, int end, ProBuilderMesh mesh)
+        {
+            if (start == lastStart && mesh == lastMesh)
+            {
+                return GetMinimalPath(lastPredecessors, start, end);
+            }
+            else
+            {
+                var predecessors = Dijkstra(start, end, mesh);
+                var path = GetMinimalPath(predecessors, start, end);
+                lastPredecessors = predecessors;
+                lastStart = start;
+                lastMesh = mesh;
+                return path;
+            }
+        }
+
+        private static int[] Dijkstra(int start, int end, ProBuilderMesh mesh)
+        {
+            HashSet<int> visited = new HashSet<int>();
+            HashSet<int> toVisit = new HashSet<int>();
+            lastWings = lastMesh == mesh ? lastWings : WingedEdge.GetWingedEdges(mesh, true);
+            int wingCount = lastWings.Count;
+
+            if (mesh != lastMesh)
+            {
+                lastFacesIndex.Clear();
+
+                for (int i = 0; i < mesh.facesInternal.Length; i++)
+                {
+                    lastFacesIndex.Add(mesh.facesInternal[i], i);
+                }
+            }
+
+            float[] weights = new float[wingCount];
+            int[] predecessors = new int[wingCount];
+
+            for (int i = 0; i < wingCount; i++)
+            {
+                weights[i] = float.MaxValue;
+            }
+
+            int current = start;
+            weights[current] = 0;
+            visited.Add(current);
+
+            // Construct the paths between the start face and every other faces
+            while (visited.Count < wingCount)
```

Comment:
Does this account for the cases where geometry is non-manifold? Ex, three faces referencing sharing a common edge, or open edges?

Reply:
It didn't but now it does!